collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py

```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
-                document.extend(self.tokenizer.text_to_ids(doc_text))
-                document.append(self.tokenizer.token_to_id("[SEP]"))
+                document = self.tokenizer.text_to_ids(doc_text)
 
-            assert len(document) >= self.max_seq_length
             return document
 
-        a_document = get_document(a_filename, a_line)
+        def match_target_seq_length(document, target_seq_length, filename,
+                                    line_idx, sentence_indices):
+            # match the target seq_length
+            for _ in range(10):
+                # if line_idx + 1 is larger than the file, then we don't match
+                # seq_length - it'll be a shorter seq_length.
+                if len(document) < target_seq_length and \
+                        line_idx + 1 < len(sentence_indices[filename]):
+                    if sentence_indices[filename][line_idx + 1] < \
+                            self.sentence_indices[filename][-1]:
+                        line_idx += 1
+                        document += get_document(
+                           filename, self.sentence_indices[filename][line_idx])
+                    else:
+                        line_idx = random.choice(
+                            range(len(sentence_indices[filename])))
+                        document = get_document(
+                            filename, sentence_indices[filename][line_idx])
+                        document, line_idx = match_target_seq_length(
+                            document, target_seq_length, filename, line_idx,
+                            sentence_indices)
+                else:
+                    break
 
-        if random.random() < 0.5:
-            # 50% of the time, B is the sentence that follows A
-            label = 1
+            return document, line_idx
 
-            a_start_idx = random.randrange(len(a_document) - seq_length)
-            b_start_idx = a_start_idx + a_length
+        target_seq_length_a = int(round(target_seq_length * 0.6))
+        target_seq_length_b = target_seq_length - target_seq_length_a
+        a_document = get_document(
+            a_filename, self.sentence_indices[a_filename][a_line_idx])
+        a_document, a_line_idx = match_target_seq_length(
+            a_document, target_seq_length_a, a_filename, a_line_idx,
+            self.sentence_indices)
 
-            b_filename = a_filename
-            b_line = a_line
-            b_document = a_document
-        else:
-            # The rest of the time, B is a random sentence from the corpus
+        if self.sentence_indices[a_filename][a_line_idx] >= \
+                self.sentence_indices[a_filename][-1] or \
+                random.random() < 0.5:
+
+            # About 50% of the time, B is a random sentence from the corpus
             label = 0
 
-            a_start_idx = random.randrange(len(a_document) - a_length)
+            for _ in range(10):
```

Comment:
same here, why is it 10?

Reply:
This is also following the original BERT implementation:
https://github.com/google-research/bert/blob/master/create_pretraining_data.py#L275

I'll add a comment.